From b1af1fc33707cb7fa2c8e07487f37e9828df83e7 Mon Sep 17 00:00:00 2001
From: Simon McVittie <smcv@debian.org>
Date: Wed, 30 Sep 2026 11:22:24 +0100
Subject: [PATCH 1/3] tests: Mock both spawn() and call() for setuptools >= 84
 compatibility
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

In older setuptools, subcommands are run via the spawn() method, which
was already replaced with a mock implementation here. In newer
setuptools, spawn() is just a deprecated wrapper around call(), and
other CCompiler methods use call() directly, so we need to patch both.

Upstream: https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/564
Helps: https://gitlab.gnome.org/GNOME/gobject-introspection/-/work_items/575
Co-authored-by: Lumír Balhar
---
 tests/scanner/test_ccompiler.py | 32 ++++++++++++++++++++++++--------
 1 file changed, 24 insertions(+), 8 deletions(-)

diff --git a/tests/scanner/test_ccompiler.py b/tests/scanner/test_ccompiler.py
index 6c0674a14..811ab2413 100644
--- a/tests/scanner/test_ccompiler.py
+++ b/tests/scanner/test_ccompiler.py
@@ -82,7 +82,10 @@ class UnixCCompilerTest(unittest.TestCase):
         except ImportError as e:
             raise unittest.SkipTest(e)
 
-        with patch.object(distutils.ccompiler.CCompiler, 'spawn') as spawn:
+        # setuptools >= 84 replaced spawn() with call(), but that might not
+        # exist in older versions
+        with patch.object(distutils.ccompiler.CCompiler, 'spawn') as spawn, \
+             patch.object(distutils.ccompiler.CCompiler, 'call', create=True) as call:
             with Environ(environ):
                 cc = CCompiler(compiler_name=compiler_name)
                 # Avoid check if target is newer from source.
@@ -90,9 +93,14 @@ class UnixCCompilerTest(unittest.TestCase):
                 # Don't actually do anything.
                 cc.compiler.dry_run = True
                 cc.compile(pkg_config_cflags, cpp_includes, [source], init_sections)
-        self.assertEqual(1, spawn.call_count)
-        args, kwargs = spawn.call_args
-        return args[0]
+
+        if spawn.call_count:
+            args, kwargs = spawn.call_args
+            return args[0]
+        else:
+            self.assertEqual(1, call.call_count)
+            args, kwargs = call.call_args
+            return args[0]
 
     def preprocess_args(self, environ={}, compiler_name=None,
                         source='a.c', output=None, cpp_options=[]):
@@ -103,7 +111,10 @@ class UnixCCompilerTest(unittest.TestCase):
         except ImportError as e:
             raise unittest.SkipTest(e)
 
-        with patch.object(distutils.ccompiler.CCompiler, 'spawn') as spawn:
+        # setuptools >= 84 replaced spawn() with call(), but that might not
+        # exist in older versions
+        with patch.object(distutils.ccompiler.CCompiler, 'spawn') as spawn, \
+             patch.object(distutils.ccompiler.CCompiler, 'call', create=True) as call:
             with Environ(environ):
                 cc = CCompiler(compiler_name=compiler_name)
                 # Avoid check if target is newer from source.
@@ -111,9 +122,14 @@ class UnixCCompilerTest(unittest.TestCase):
                 # Don't actually do anything.
                 cc.compiler.dry_run = True
                 cc.preprocess(source, output, cpp_options)
-        self.assertEqual(1, spawn.call_count)
-        args, kwargs = spawn.call_args
-        return args[0]
+
+        if spawn.call_count:
+            args, kwargs = spawn.call_args
+            return args[0]
+        else:
+            self.assertEqual(1, call.call_count)
+            args, kwargs = call.call_args
+            return args[0]
 
     @unittest.skip("Currently a Python build time compiler is used as the default.")
     def test_compile_default(self):
-- 
GitLab
